rct229/rules/section5.py

```diff
@@ -110,6 +116,114 @@ def rule_check(self, context, calc_vals, data=None):
 # ------------------------
 
 
+class Section5Rule5(RuleDefinitionListIndexedBase):
+    """Rule 5 of ASHRAE 90.1-2019 Appendix G Section 5 (Envelope)"""
+
+    def __init__(self):
+        super(Section5Rule5, self).__init__(
+            rmrs_used=UserBaselineProposedVals(False, True, False),
+            required_fields={
+                "weather": ["climate_zone"],
+            },
+            each_rule=Section5Rule3.BuildingRule(),
+            index_rmr="baseline",
+            id="5-5",
+            description="Baseline roof assemblies must match the appropriate assembly maximum U-factors in Tables G3.4-1 through G3.4-8.",
+            list_path="buildings[*]",
+        )
+
+        def create_data(context, data=None):
+            rmr_baseline = context.baseline
+            return {"climate_zone": rmr_baseline["weather"]["climate_zone"]}
+
+    class BuildingRule(RuleDefinitionListIndexedBase):
+        def __init__(self):
+            super(Section5Rule5.BuildingRule, self).__init__(
+                rmrs_used=UserBaselineProposedVals(False, True, False),
+                required_fields={},
+                each_rule=Section5Rule3.BuildingRule.RoofRule(),
+                index_rmr="baseline",
+            )
+
+        def create_context_list(context, data=None):
+            building = context.baseline
+            # List of all baseline roof surfaces to become the context for RoofRule
+            return [
+                UserBaselineProposedVals(None, surface, None)
+                for surface in find_all("$..surfaces[*]", building)
+                if get_opaque_surface_type(surface) == ROOF
+            ]
+
+        def create_data(self, context, data=None):
+            building = context.baseline
+            # Merge into the existing data dict
+            return {
+                **data,
+                surface_conditioning_category_dict: get_surface_conditioning_category_dict(
+                    data["climate_zone"], building
+                ),
+            }
+
+        class RoofRule(RuleDefinitionBase):
+            def __init__(self):
+                super(Section5Rule5.BuildingRule.RoofRule, self).__init__(
+                    rmrs_used=UserBaselineProposedVals(False, True, False),
+                    required_fields={},
+                )
+
+            def get_calc_vals(self, context, data=None):
+                climate_zone: str = data["climate_zone"]
```

Comment:
I like the way it defines the variable types and wondering whether we should enforce the style in this project? Will it cause a lot overheads?

Reply:
Yes, I have started to do this with new code – at least for the few variables that are primitive types. I need to learn more about how to handle more complex types.